### PR TITLE
Working

### LINKED ISSUE #1: Future
## Future

Define a schema for the directory.

``` ruby
class ImageDir
  include DirModel

  file :zone_image, path: "{{root}}/{{branch}}", name: "{{zone_name}}" 
end
```

To export, define your export model like [`ActiveModel::Serializer`](https://github.com/rails-api/active_model_serializers)
and generate the directory requirements:

``` ruby
class ImageExportDir < ImageDir
  include DirModel::Export

  # below are method overrides with default implementation
  #
  # overriding is recommended
  #
  def root; source_model.root end
  def branch; source_model.branch end
  def zone_name; source_model.zone_name end

  # define your image source
  #
  # override with something like: File.new("path/to/image.png")
  #
  def zone_image
    source_model.zone_image
  end
end

dir = DirModel::Export::AggregateDir.new(ImageExportDir, context).generate do |dir|
   dir << source_model
end

#then you can:
zipper = ZipDir::Zipper.new
zip_file = zipper.generate(dir.path)

# and `dir_model` will be independent of  `zip_dir` (no need to check `require` order
```

To import, define your import model, which works like [`ActiveRecord`](http://guides.rubyonrails.org/active_record_querying.html)
wrapping over a directory:

``` ruby
class ImageImportDir < ImageDir
  include DirModel::Import

  #
  # below are method overrides with default implementation
  #
  def root; "level1" end
  def branch; "level2" end
  def zone_name; "testing" end
  def zone_image; File.new("....") end

  presenter do
     # presenter magic
  end
end

DirModel::Import::AggregateDir.new(path, ImageImportDir, context).each do |dir_model|
   # dir_model.....
end
```

### PR DESCRIPTION
**Pull request by joel**
closes #1

### CODE REVIEW

File: .ruby-version [hunk 1]
```diff
@@ -0,0 +1 @@
+2.2.3
```
Comment (s12chung) [inline .ruby-version]: https://github.com/FinalCAD/inherited_class_var/pull/1#discussion_r43904163

File: lib/dir_model/export/aggregate_dir.rb [hunk 2]
```diff
@@ -0,0 +1,25 @@
+require 'active_support/hash_with_indifferent_access'
```
Comment (s12chung) [inline lib/dir_model/export/aggregate_dir.rb:1]: not needed?

Comment (s12chung) [inline lib/dir_model/export/aggregate_dir.rb:1]: ah... `symbolize_keys`, might be a good idea?

File: lib/dir_model/export/aggregate_dir.rb [hunk 3]
```diff
@@ -0,0 +1,25 @@
...
+  def initialize(export_dir_model_class, context={})
+    @export_dir_model_class = export_dir_model_class
+    @context = context.to_h #.symbolize_keys
+  end
+
+  # Add a row_model to the
+  # @param [] source_model the source model of the export file model
+  # @param [Hash] context the extra context given to the instance of the file model
+  def append_model(source_model, context={})
+    @paths << export_dir_model_class.new(source_model, context.to_h.reverse_merge(self.context)).path
```
Comment (s12chung) [inline lib/dir_model/export/aggregate_dir.rb]: my bad: https://github.com/FinalCAD/csv_row_model/commit/5171808bebb0682e73e5e71e237faeb603ef216b

File: lib/dir_model/export/aggregate_dir.rb [hunk 4]
```diff
@@ -0,0 +1,25 @@
+require 'active_support/hash_with_indifferent_access'
+
+class DirModel::Export::AggregateDir
```
Comment (s12chung) [inline lib/dir_model/export/aggregate_dir.rb]: think it's better convention to have multiple `module` calls....

File: spec/fixtures/models/image_dir.rb [hunk 5]
```diff
@@ -0,0 +1,5 @@
+class ImageDir
+  include DirModel
+
+  file :my_image, path: "{{root}}/{{dir}}/{{sub_dir}}", name: :my_image_name
```
Comment (s12chung) [inline spec/fixtures/models/image_dir.rb]: :warning:  btw, what are you thoughts on this syntax?

``` ruby
file :my_image, path: -> { "#{root}/#{dir}/#{sub_dir}" }, name: -> { "my_#{image}_name" }
```

pro: no new template language, uses natural ruby instead.
con: interpolating the methods available might be complicated (but possible, with some magic)

File: lib/dir_model/utils.rb [hunk 6]
```diff
@@ -0,0 +1,7 @@
+module Utils
+  extend ActiveSupport::Concern
+
+  def mkdir
+    FileUtils.mkdir_p(yield)
+  end
+end
```
Comment (s12chung) [inline lib/dir_model/utils.rb:7]: only used once. no need to extract imo.

File: spec/fixtures/models/image_export_dir.rb [hunk 7]
```diff
@@ -0,0 +1,23 @@
+class ImageExportDir < ImageDir
+  include DirModel::Export
+
+  def root
+    "tmp/root_dir_#{source_model.id}"
```
Comment (s12chung) [inline spec/fixtures/models/image_export_dir.rb]: let's remove the root part---think it'll confuse other readers.

Comment (s12chung) [inline spec/fixtures/models/image_export_dir.rb]: (because of the `tmp`)

File: lib/dir_model/export.rb [hunk 8]
```diff
@@ -1,82 +1,49 @@
...
+
+        File.open(File.join(@root_path, file_path), 'wb') {|f| f.write(self.public_send(file_name).read) }
+      end
     ensure
       @generated = true
     end
 
     def cleanup
-      FileUtils.remove_entry_secure @temp_path
+      entry_paths.each { |file_path| FileUtils.remove_entry_secure file_path }
```
Comment (s12chung) [inline lib/dir_model/export.rb]: simplification without loops:

``` ruby
FileUtils.remove_entry_secure @root_path

# under cleanup if generated? above, remove from constructor
@root_path    = Dir.mktmpdir
```

File: lib/dir_model/export.rb [hunk 9]
```diff
@@ -1,82 +1,49 @@
...
+      entry_paths
     end
-    def file_name(file)
-      File.join(*pwd, public_send(self.class.file_name_method_name(file)))
+
+    private
+
+    def entry_paths
+      Dir.clean_entries(@root_path).map { |entry| File.join(@root_path, entry) }
     end
```
Comment (s12chung) [inline lib/dir_model/export.rb]: let's keep the interface similar, just so it's easier to debug with 1 path?

```
public
-    def path       
-      generate unless generated?       
-      @root_path       
-    end

   def entry_paths
        generate unless generated?
+      Dir.clean_entries(@root_path).map { |entry| File.join(@root_path, entry) }
     end
```

File: lib/dir_model/export.rb [hunk 10]
```diff
@@ -1,82 +1,49 @@
...
+
     def generate
       cleanup if generated?
-      @temp_path = Dir.mktmpdir
 
-      @pwd = [@temp_path]
-      _generate
+      self.class.files.each do |file_name, options|
+        dir_path  = instance_exec(&options[:path])
+        file_path = File.join(dir_path, instance_exec(&options[:name]))
```
Comment (s12chung) [inline lib/dir_model/export.rb:30]: would be nice to be able to accept string for the options. can extract this out into two methods.

File: spec/fixtures/models/image_model.rb [hunk 11]
```diff
@@ -0,0 +1,18 @@
...
+  end
+
+  def zone_name
+    'zone_name'
+  end
+
+  def zone
+    File.new('spec/fixtures/image.png')
+  end
+end
```
Comment (s12chung) [inline spec/fixtures/models/image_model.rb]: let's delete this file and replace it with an `openstruct`

File: spec/dir_model/export/aggregate_dir_spec.rb [hunk 12]
```diff
@@ -0,0 +1,17 @@
+require 'spec_helper'
+
+describe DirModel::Export::AggregateDir do
+  let(:file_contents)    { File.new('spec/fixtures/image.png').read }
+  let(:source_model)     { ModelImage.new }
+  let(:instance)         { DirModel::Export::AggregateDir.new(ImageExportDir) }
+  let(:file_path)        { '/tmp/root_dir_42/Sectors/sector_name/zone_name.png' }
```
Comment (s12chung) [inline spec/dir_model/export/aggregate_dir_spec.rb]: let's have > 1 source models

**Comment by s12chung** [comment]
overall. looks good :white_check_mark: --- all of the above are minor polish. let's add some tests for `Export`?

File: lib/dir_model/export/aggregate_dir.rb [hunk 13]
```diff
@@ -0,0 +1,40 @@
...
+        @dir_path              = Dir.mktmpdir
+      end
+
+      # Add a row_model to the
+      # @param [] source_model the source model of the export file model
+      # @param [Hash] context the extra context given to the instance of the file model
+      def append_model(source_model, context={})
+        export_dir_model_class.new(source_model, context.reverse_merge(self.context)).paths.each do |file_path|
+          add_path(file_path)
+        end
```
Comment (s12chung) [inline lib/dir_model/export/aggregate_dir.rb]: i just remembered. let's do a `cleanup` call here.

File: lib/dir_model/utils.rb (hunk 6 above)
Comment (joel) [inline lib/dir_model/utils.rb:7]: Was twice, but yeah isn't necessary now

File: lib/dir_model/export.rb (hunk 10 above)
Comment (joel) [inline lib/dir_model/export.rb:30]: I'm not sure to understand here

File: spec/fixtures/models/image_model.rb (hunk 11 above)
Comment (joel) [inline spec/fixtures/models/image_model.rb]: I don't understand what you want to do here

File: lib/dir_model/export.rb (hunk 10 above)
Comment (joel) [inline lib/dir_model/export.rb:30]: This ? https://github.com/FinalCAD/dir_model/commit/aa3b81411c79facd64742d31ace06851ba281dd6

Comment (s12chung) [inline lib/dir_model/export.rb:30]:
``` ruby
class DirModel
   file name: "some_file_name_without_proc", path: "some_path/without/proc"
end
```

slightly cleaner syntax that we can actually use.

**Pull request merged**